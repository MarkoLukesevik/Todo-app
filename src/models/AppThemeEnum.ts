export enum AppThemeEnum {
    Light = 0,
    Dark = 1
}
