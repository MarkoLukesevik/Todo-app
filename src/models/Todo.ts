export interface Todo {
  id: string
  title: string
  isComplete: false
}
