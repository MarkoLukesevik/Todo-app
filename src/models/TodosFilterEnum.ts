export enum TodosFilterEnum {
    All = 0,
    Active = 1,
    Completed = 2
}