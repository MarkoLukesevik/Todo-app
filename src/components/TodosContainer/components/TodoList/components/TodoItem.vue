<script setup lang="ts">
import type { Todo } from '@/models/Todo';
import CheckButton from '@/base-components/CheckButton.vue';
import XButton from '@/base-components/XButton.vue';
import { AppThemeEnum } from '@/models/AppThemeEnum';
import { computed } from 'vue';

const props = defineProps<{
    appTheme: AppThemeEnum;
    todo: Todo;
}>();

const emit =defineEmits<{
    (e: 'toggle-todo', value: string): void;
    (e: 'delete-todo', value: string): void;
}>();

const todoItemClasses = computed<string[]>(() => {
    const classes = ['todo-item'];

    if (props.appTheme === AppThemeEnum.Dark)
        classes.push('todo-item__dark')

    return classes
});
</script>

<template>
    <div :class="todoItemClasses">
        <div class="todo-item-content">
            <check-button
                :isChecked="todo.isComplete"
                :app-theme="appTheme"
                @toggle="emit('toggle-todo', todo.id)"
            />
            <p>{{ todo.title }}</p>
        </div>
        <x-button @button-click="emit('delete-todo', todo.id)"></x-button>
    </div>
</template>

<style scoped lang="scss">
.todo-item {
    display: flex;
    flex-direction: row;
    flex: 0 1 auto;
    align-items: center;
    justify-content: space-between;
    column-gap: 0.5rem;

    color: #494C6B;
    border-bottom: 1px solid #E3E4F1;
    padding: 1rem 1.2rem;

    font-family: 'Josefin Sans', sans-serif;
    font-size: 0.9rem;
    font-weight: 400;

     &__dark {
        border-bottom: 1px solid #393A4B;
        color: #C8CBE7;
    }

    &-content {
        display: flex;
        flex-direction: row;
        flex: 0 1 auto;
        align-items: center;
        column-gap: 0.7rem;
    }
}

@media only screen and (min-width: 768px) {
    .todo-item {
        font-size: 1.1rem;
    }
}
</style>
