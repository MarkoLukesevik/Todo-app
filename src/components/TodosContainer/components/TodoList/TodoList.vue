<script setup lang="ts">
import { computed, ref, watch, type Ref } from 'vue';

import draggable from 'vuedraggable';

import type { Todo } from '@/models/Todo';
import { AppThemeEnum } from '@/models/AppThemeEnum';
import { TodosFilterEnum } from '@/models/TodosFilterEnum';

import TodoItem from './components/TodoItem.vue';

const props = defineProps<{
    appTheme:AppThemeEnum;
    todos: Todo[];
    todosFilter: TodosFilterEnum;
    todosCount: number;
}>();

const scopedTodos: Ref<Todo[]> = ref(props.todos);

watch(() => props.todos, () => {
    scopedTodos.value = props.todos;
});

const emit = defineEmits<{
    (e: 'toggle-todo', value: string): void;
    (e: 'delete-todo', value: string): void;
    (e: 'delete-completed-todos'): void;
    (e: 'change-todos-filter', value: TodosFilterEnum): void;
    (e: 'update:model-value', value: Todo[]): void;
}>();

const todosClasses = computed<string[]>(() => {
    const classes = ['todos'];

    if (props.appTheme === AppThemeEnum.Dark)
        classes.push('todos-dark');

    return classes;
});
</script>

<template>
    <div :class="todosClasses">
        <p v-if="todosCount === 0" class="todos-empty">Your todo list is empty. please add some todo items.</p>
        <div class="todos-list">
            <draggable
                v-model="scopedTodos"
                :item-key="'todos'"
                :delay="100"
                @change=" emit('update:model-value', scopedTodos)"
            >
                <template #item="{ element: todo }">
                    <todo-item
                        :key="todo.id"
                        :app-theme="appTheme"
                        :todo="todo"
                        @toggle-todo="emit('toggle-todo', $event)"
                        @delete-todo="emit('delete-todo', $event)"
                    />
                </template>
            </draggable>
        </div>
        <div class="todos-footer">
            <div class="todos-footer__counter">{{ todosCount }} items left</div>
            <div class="todos-footer__actions">
                <span
                    :class="{ 'active-filter': todosFilter === TodosFilterEnum.All }"
                    @click="emit('change-todos-filter', TodosFilterEnum.All)"
                >
                    All
                </span>
                <span
                    :class="{ 'active-filter': todosFilter === TodosFilterEnum.Active }"
                    @click="emit('change-todos-filter', TodosFilterEnum.Active)"
                >
                    Active
                </span>
                <span
                    :class="{ 'active-filter': todosFilter === TodosFilterEnum.Completed }"
                    @click="emit('change-todos-filter', TodosFilterEnum.Completed)"
                >
                    Completed
                </span>
            </div>
            <div class="todos-footer__clear" @click="emit('delete-completed-todos')">Clear Completed</div>
        </div>
    </div>
</template>

<style scoped lang="scss">
.todos {
    display: flex;
    flex-direction: column;
    flex: 1 1 auto;
        
    background-color: #ffffff;
    border-radius: 5px;
    box-shadow: 0px 35px 50px -15px rgba(194, 195, 214, 0.5);

    ::-webkit-scrollbar {
        width: 0.3rem;
    }

    ::-webkit-scrollbar-track {
            background-color: #ffffff;
        }

    ::-webkit-scrollbar-thumb {
        background-color: #E3E4F1;
        border-radius: 5px;
    }

    &-empty {
        margin: 1rem 1.2rem;
        color: #494C6B;
        font-family: 'Josefin Sans', sans-serif;
        font-size: 0.9rem;
        font-weight: 400;
    }

    &-list {
        display: flex;
        flex-direction: column;
        flex: 1 1 0;

        overflow-y: auto;
    }

    &-footer {
        display: flex;
        flex-direction: row;
        flex: 0 1 auto;
        align-items: center;
        justify-content: space-between;
        column-gap: 0.5rem;

        padding: 1rem 1.2rem;
        color: #9495A5;
    
        font-family: 'Josefin Sans', sans-serif;
        font-size: 0.9rem;
        font-weight: 400;

        &__actions {
            display: none;
        }

        &__clear {
            cursor: pointer;

            &:hover {
                color: #494C6B;
            }
        }
    }

    &-dark {
         background-color: #25273D;
         box-shadow: 0px 35px 50px -15px rgba(0, 0, 0, 0.5);

         & .todos-empty {
            color: #C8CBE7;
         }

         & .todos-footer {
            color: #5B5E7E;

            &__clear:hover {
                color: #E3E4F1;
            }
         }

        ::-webkit-scrollbar-track {
            background-color: #25273D;
        }

        ::-webkit-scrollbar-thumb {
            background-color: #393A4B;
            border-radius: 5px;
        }
    }
}

@media only screen and (min-width: 768px) {
    .todos {
        &-empty {
            font-size: 1.1rem;
        }

        &-footer {
            font-size: 1rem;

            &__actions {
                display: flex;
                flex-direction: row;
                flex: 0 1 auto;
                align-items: center;
                justify-content: center;
                column-gap: 1rem;

                & span {
                    cursor: pointer;
                    font-weight: 600;

                    &:hover {
                        color: #494C6B;
                    }
                }

                & .active-filter {
                    color: #3A7CFD;
                }
            }
        }
    }

    .todos-dark .todos-footer__actions span:hover {
        color: #E3E4F1;
    }
}
</style>