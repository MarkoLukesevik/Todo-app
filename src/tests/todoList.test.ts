import { describe, it, expect, beforeEach } from 'vitest';
import { shallowMount } from '@vue/test-utils';

import { AppThemeEnum } from '@/models/AppThemeEnum';
import { TodosFilterEnum } from '@/models/TodosFilterEnum';
import type { Todo } from '@/models/Todo';

import TodoList from '@/components/TodosContainer/components/TodoList/TodoList.vue';
import TodoItem from '@/components/TodosContainer/components/TodoList/components/TodoItem.vue';

describe('Todos', () => {
  let wrapper: ReturnType<typeof shallowMount>;
  const todos: Todo[] = [
    { id: '1', title: 'First Todo', isComplete: false },
    { id: '2', title: 'Second Todo', isComplete: true },
  ];

    beforeEach(() => {
        wrapper = shallowMount(TodoList, {
            props: {
                appTheme: AppThemeEnum.Light,
                todos,
                todosFilter: TodosFilterEnum.All,
                todosCount: todos.length,
            },
            global: {
                components: {
                TodoItem: TodoItem,
                draggable: { template: '<div />' },
                },
            },
        });
    });

    it('Should match the snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('Should render correctly with default props', () => {
        const todosDiv = wrapper.find('[data-test="todos-list"]');

        expect(todosDiv.exists()).toBe(true);
        expect(todosDiv.classes()).toContain('todos');
        expect(wrapper.find('[data-test="todos-list-empty"]').exists()).toBe(false);
        expect(wrapper.find('[data-test="todos-list-counter"]').text()).toBe('2 items left');
    });

    it('Should display empty message when todosCount is 0', async () => {
        await wrapper.setProps({ todosCount: 0 });

        expect(wrapper.find('[data-test="todos-list-empty"]').exists()).toBe(true);
        expect(wrapper.find('[data-test="todos-list-counter"]').text()).toBe('0 items left');
    });

    it('Should apply dark theme class when appTheme is Dark', async () => {
        await wrapper.setProps({ appTheme: AppThemeEnum.Dark });
        expect(wrapper.find('.todos').classes()).toContain('todos-dark');
    });

    it('Should emit correct events when filter options are clicked', async () => {
        const allFilter = await wrapper.find('[data-test="todos-list-all-filter"]');
        const activeFilter = await wrapper.find('[data-test="todos-list-active-filter"]');
        const completedFilter = await wrapper.find('[data-test="todos-list-completed-filter"]');

        await allFilter.trigger('click');
        expect(wrapper.emitted('change-todos-filter')).toBeTruthy();
        expect(wrapper.emitted('change-todos-filter')?.[0]).toEqual([TodosFilterEnum.All]);

        await activeFilter.trigger('click');
        expect(wrapper.emitted('change-todos-filter')?.[1]).toEqual([TodosFilterEnum.Active]);

        await completedFilter.trigger('click');
        expect(wrapper.emitted('change-todos-filter')?.[2]).toEqual([TodosFilterEnum.Completed]);
    });

    it('Should emit delete-completed-todos event when Clear Completed is clicked', async () => {
        await wrapper.find('[data-test="todos-list-clear"]').trigger('click');
        expect(wrapper.emitted('delete-completed-todos')).toBeTruthy();
    });

    it('Should display correct filter classes based on selected filter', async () => {
        await wrapper.setProps({ todosFilter: TodosFilterEnum.All });
        expect(wrapper.find('[data-test="todos-list-all-filter"]').classes()).toContain('active-filter');
        
        await wrapper.setProps({ todosFilter: TodosFilterEnum.Active });
        expect(wrapper.find('[data-test="todos-list-active-filter"]').classes()).toContain('active-filter');

        await wrapper.setProps({ todosFilter: TodosFilterEnum.Completed });
        expect(wrapper.find('[data-test="todos-list-completed-filter"]').classes()).toContain('active-filter');
    });
});