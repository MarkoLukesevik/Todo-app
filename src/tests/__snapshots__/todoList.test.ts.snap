// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todos > Should match the snapshot 1`] = `
<div
  class="todos"
  data-test="todos-list"
  data-v-fbbc3ae1=""
>
  <!--v-if-->
  <div
    class="todos-list"
    data-v-fbbc3ae1=""
  >
    <draggable-stub
      clone="[Function]"
      data-v-fbbc3ae1=""
      delay="1"
      disabled="false"
      itemkey="todos"
      modelvalue="[object Object],[object Object]"
      tag="div"
    />
  </div>
  <div
    class="todos-footer"
    data-v-fbbc3ae1=""
  >
    <div
      class="todos-footer__counter"
      data-test="todos-list-counter"
      data-v-fbbc3ae1=""
    >
      2 items left
    </div>
    <div
      class="todos-footer__actions"
      data-v-fbbc3ae1=""
    >
      <span
        class="active-filter"
        data-test="todos-list-all-filter"
        data-v-fbbc3ae1=""
      >
         All 
      </span>
      <span
        class=""
        data-test="todos-list-active-filter"
        data-v-fbbc3ae1=""
      >
         Active 
      </span>
      <span
        class=""
        data-test="todos-list-completed-filter"
        data-v-fbbc3ae1=""
      >
         Completed 
      </span>
    </div>
    <div
      class="todos-footer__clear"
      data-test="todos-list-clear"
      data-v-fbbc3ae1=""
    >
       Clear Completed 
    </div>
  </div>
</div>
`;
